client.go

```diff
@@ -0,0 +1,593 @@
+	default:
+		return nil, fmt.Errorf("action read: unexpected code: %v", msg.Code)
+	}
+
+	return act, nil
+}
+
+// Helo sends the connection information to the milter.
+//
+// It should be called once per milter session (from NewSession to Close).
+func (s *ClientSession) Conn(hostname string, family ProtoFamily, port uint16, addr string) (*Action, error) {
+	if s.protoMask&OptNoConnect != 0 {
+		return &Action{Code: ActContinue}, nil
+	}
+
+	msg := &Message{
+		Code: byte(CodeConn),
+	}
+	msg.Data = appendCString(msg.Data, hostname)
+	msg.Data = append(msg.Data, byte(family))
+	if family != FamilyUnknown {
+		if family == FamilyInet || family == FamilyInet6 {
+			msg.Data = appendUint16(msg.Data, port)
+		}
+		msg.Data = appendCString(msg.Data, addr)
+	}
+
+	if err := writePacket(s.conn, msg, s.writeTimeout); err != nil {
+		return nil, fmt.Errorf("milter: conn: %w", err)
+	}
+
+	act, err := s.readAction()
+	if err != nil {
+		return nil, fmt.Errorf("milter: conn: %w", err)
+	}
+	return act, nil
+}
+
+// Helo sends the HELO hostname to the milter.
+//
+// It should be called once per milter session (from NewSession to Close).
+func (s *ClientSession) Helo(helo string) (*Action, error) {
+	// Synthesise response as if server replied "go on" while in fact it does
+	// not support that message.
+	if s.protoMask&OptNoHelo != 0 {
+		return &Action{Code: ActContinue}, nil
+	}
+
+	msg := &Message{
+		Code: byte(CodeHelo),
+		Data: appendCString(nil, helo),
+	}
+
+	if err := writePacket(s.conn, msg, s.writeTimeout); err != nil {
+		return nil, fmt.Errorf("milter: helo: %w", err)
+	}
+
+	act, err := s.readAction()
+	if err != nil {
+		return nil, fmt.Errorf("milter: helo: %w", err)
+	}
+	return act, nil
+}
+
+func (s *ClientSession) Mail(sender string, esmtpArgs []string) (*Action, error) {
+	if s.protoMask&OptNoMailFrom != 0 {
+		return &Action{Code: ActContinue}, nil
+	}
+
+	msg := &Message{
+		Code: byte(CodeMail),
+	}
+
+	msg.Data = appendCString(msg.Data, "<"+sender+">")
+	for _, arg := range esmtpArgs {
+		msg.Data = appendCString(msg.Data, arg)
+	}
+
+	if err := writePacket(s.conn, msg, s.writeTimeout); err != nil {
+		return nil, fmt.Errorf("milter: mail: %w", err)
+	}
+
+	act, err := s.readAction()
+	if err != nil {
+		return nil, fmt.Errorf("milter: mail: %w", err)
+	}
+	return act, nil
+}
+
+func (s *ClientSession) Rcpt(rcpt string, esmtpArgs []string) (*Action, error) {
+	if s.protoMask&OptNoRcptTo != 0 {
+		return &Action{Code: ActContinue}, nil
+	}
+
+	msg := &Message{
+		Code: byte(CodeRcpt),
+	}
+
+	msg.Data = appendCString(msg.Data, "<"+rcpt+">")
+	for _, arg := range esmtpArgs {
+		msg.Data = appendCString(msg.Data, arg)
+	}
+
+	if err := writePacket(s.conn, msg, s.writeTimeout); err != nil {
+		return nil, fmt.Errorf("milter: rcpt: %w", err)
+	}
+
+	act, err := s.readAction()
+	if err != nil {
+		return nil, fmt.Errorf("milter: rcpt: %w", err)
+	}
+	return act, nil
+}
+
+// HeaderField sends a single header field to the milter.
```

Comment:
Does this expect raw fields (with whitespace preserved)?

Reply:
Protocol does not seem to specify this so I assume it is (think of OpenDKIM milter).

Reply:
Documented the raw fields requirement in 563ba50.